### PR TITLE
Master worker deployment

### PR DESCRIPTION
**Pull request by olga-mir**
## Goal:
Networking settings are different for master and workers. We want to have 2 internet checker controllers running at the same time - one on a master node and one on a worker node.

## Implementation:
We investigated 2 approaches: One deployment with 2 pods spread out with podAntiAffinity and 2 separate deployments. 

### One deployment with 2 pods and anti affinity rules
- it is possible to guarantee the pod spread on master vs node using podAntiAfinity *if* we add a new label to nodes (it requires same label with different values on master/node, currently there is no such label. The role label is of form: `node-role.kubernetes.io/master: ""`, which doesn't help to achieve the goal)
- We will be forced to use update strategy `Recreate` as RollingUpdate does not work in this scenario - the pods lock themselves out and the deployment stuck. `recreate` means some downtime, which is not that critical at this point maybe with an exception of alertmanager webhook sink. 

### Two deployments that explicitly target master/worker node
This PR implements this approach. 
Additional advantage is that worker pod has minimal privileges (no scc, access to secrets, etc)

### CODE REVIEW

**Comment by olga-mir** [comment]
```
$ oc get pods -o wide
NAME                                  READY   STATUS    RESTARTS   AGE   IP            NODE                                            NOMINATED NODE   READINESS GATES
aro-operator-master-7c9d8c795-kj54g   1/1     Running   0          24m   10.128.0.24   olga-v4-6g86s-master-1                          <none>           <none>
aro-operator-worker-cdd7f66bc-sbq9w   1/1     Running   0          24m   10.131.0.12   olga-v4-6g86s-worker-australiasoutheast-59f42   <none>           <none>
```

```
$ get cluster.aro.openshift.io cluster -o yaml 
  conditions:
  - lastTransitionTime: "2020-06-02T01:24:55Z"
    message: |
      [https://management.azure.com/subscriptions/<SUB_ID>/resourceGroups/v4-australiasoutheast/providers/Microsoft.RedHatOpenShift/openShiftClusters/olga-v4?api-version=2020-04-30] Get https://management.azure.com/subscriptions/<SUB_ID>/resourceGroups/v4-australiasoutheast/providers/Microsoft.RedHatOpenShift/openShiftClusters/olga-v4?api-version=2020-04-30: dial tcp: lookup management.azure.com on 172.30.0.10:53: no such host
    reason: CheckFailed
    status: "False"
    type: InternetReachableFromMaster
  - lastTransitionTime: "2020-06-02T01:01:00Z"
    message: |
      [https://management.azure.com/subscriptions/<SUB_ID>/resourceGroups/v4-australiasoutheast/providers/Microsoft.RedHatOpenShift/openShiftClusters/olga-v4?api-version=2020-04-30] Get https://management.azure.com/subscriptions/<SUB_ID>/resourceGroups/v4-australiasoutheast/providers/Microsoft.RedHatOpenShift/openShiftClusters/olga-v4?api-version=2020-04-30: dial tcp: lookup management.azure.com on 172.30.0.10:53: no such host
    reason: CheckFailed
    status: "False"
    type: InternetReachableFromWorker
```

File: cmd/aro/operator.go [hunk 1]
```diff
@@ -76,54 +78,93 @@ func operator(ctx context.Context, log *logrus.Entry) error {
...
+
 	// +kubebuilder:scaffold:builder
 
 	log.Info("starting manager")
 	return mgr.Start(ctrl.SetupSignalHandler())
 }
+
+// Parses `nodeName` environment variable to retrieve pod's placement - master or worker
+func getPlacement(log *logrus.Entry) (string, error) {
+	nodeName, found := os.LookupEnv("NODE_NAME")
```
Comment (asalkeld) [inline cmd/aro/operator.go:149]: maybe also cater for "go run ./cmd/aro operator" - if RP_MODE=development then assume master

File: operator/deploy/deploy.go [hunk 2]
```diff
@@ -232,6 +222,38 @@ func (o *operator) deployment() *appsv1.Deployment {
...
+				Name: "pullsecret-tokens",
+				VolumeSource: v1.VolumeSource{
+					Secret: &v1.SecretVolumeSource{
+						SecretName: "pullsecret-tokens",
+					},
+				},
+			},
+		}
+		deployment.Spec.Template.Spec.NodeSelector = map[string]string{"node-role.kubernetes.io/master": ""}
+		deployment.Spec.Template.Spec.Tolerations = []v1.Toleration{
```
Comment (asalkeld) [inline operator/deploy/deploy.go:237]: you are not setting this stuff for worker? I assume it defaults to non-master?

File: operator/deploy/deploy.go [hunk 3]
```diff
@@ -345,12 +368,20 @@ func (o *operator) CreateOrUpdate(ctx context.Context) error {
...
-		return false, err
+	dm, errMaster := o.cli.AppsV1().Deployments(o.namespace).Get("aro-operator-master", metav1.GetOptions{})
+	if errMaster != nil {
+		return false, errMaster
+	}
+
+	dw, errWorker := o.cli.AppsV1().Deployments(o.namespace).Get("aro-operator-worker", metav1.GetOptions{})
+	if errWorker != nil {
+		return false, errWorker
 	}
```
Comment (asalkeld) [inline operator/deploy/deploy.go:379]: you could just use "err" in both cases

File: operator/deploy/deploy.go (hunk 2 above)
Comment (olga-mir) [reply operator/deploy/deploy.go:237]: I will setup NodeSelector explicitly for the worker. But generally if the deployment doesn't have specific toleration set, then it get scheduled on the nodes without that toleration. This ensures worker don't get scheduled on a master

**Comment by olga-mir** [comment]
closing in favor of https://github.com/Azure/ARO-RP/pull/791

**Pull request closed without merging**